source/common/http/filter_manager.cc

```diff
@@ -183,11 +183,12 @@ bool ActiveStreamFilterBase::commonHandleAfterDataCallback(FilterDataStatus stat
       buffer_was_streaming = status == FilterDataStatus::StopIterationAndWatermark;
       commonHandleBufferData(provided_data);
     } else if (complete() && !hasTrailers() && !bufferedData()) {
-      // If this filter is doing StopIterationNoBuffer and this stream is terminated with a zero
-      // byte data frame, we need to create an empty buffer to make sure that when commonContinue
-      // is called, the pipeline resumes with an empty data frame with end_stream = true
-      ASSERT(end_stream_);
-      bufferedData() = createBuffer();
+      if (end_stream_) {
+        // If this filter is doing StopIterationNoBuffer and this stream is terminated with a zero
+        // byte data frame, we need to create an empty buffer to make sure that when commonContinue
+        // is called, the pipeline resumes with an empty data frame with end_stream = true
+        bufferedData() = createBuffer();
+      }
     }
```

Comment:
TODO: This ASSERT was introduced in https://github.com/envoyproxy/envoy/pull/7720. I need to check why. The assertion fails when we `sendLocalReply` in `encodeData` for a server streaming case.

Reply:
This change is part of #15075 now. Feel free to ignore the diff here.